include/envoy/config/typed_metadata.h

```diff
@@ -0,0 +1,58 @@
+#pragma once
+
+#include <string>
+
+#include "envoy/stream_info/filter_state.h"
+
+#include "common/protobuf/protobuf.h"
+
+namespace Envoy {
+namespace Config {
+
+/**
+ * TypedMetadata interface.
+ */
+class TypedMetadata {
+public:
+  virtual ~TypedMetadata() {}
+
+  // Returns a T instance by key. If the conversion is not able to complete, or
+  // if the data is not in the store, returns a nullptr.
+  template <typename T> const T* get(const std::string& key) const {
+    static_assert(std::is_base_of<StreamInfo::FilterState::Object, T>::value,
+                  "Data type must be subclass of StreamInfo::FilterState::Object");
+    const StreamInfo::FilterState::Object* p = getData(key);
+    if (p != nullptr) {
+      return dynamic_cast<const T*>(p);
+    }
+    return nullptr;
+  }
+
+protected:
+  // Returns data associated with given 'key'.
```

Comment:
Can you add a fuller Doxygen comment? In particular, what happens if this is `nullptr`? Or should it be an `Object&`?

Reply:
.. and also do the `/** ... */` Doxygen style docs. Need to also have a `@return`. Use https://github.com/envoyproxy/envoy/blob/master/include/envoy/server/admin.h#L126 as an example for what these should look like.

Reply:
oh lord, doxygen.